crates/store/src/state.rs

```diff
@@ -698,45 +703,80 @@ impl State {
+
         let state_headers = if !include_headers {
             BTreeMap::<AccountId, AccountStateHeader>::default()
         } else {
             let infos = self.db.select_accounts_by_ids(account_ids.clone()).await?;
-
             if account_ids.len() > infos.len() {
                 let found_ids = infos.iter().map(|info| info.summary.account_id).collect();
                 return Err(DatabaseError::AccountsNotFoundInDb(
                     BTreeSet::from_iter(account_ids).difference(&found_ids).copied().collect(),
                 ));
             }
 
-            infos
-                .into_iter()
-                .filter_map(|info| {
-                    info.details.map(|details| {
-                        (
-                            info.summary.account_id,
-                            AccountStateHeader {
-                                header: Some(AccountHeader::from(&details).into()),
-                                storage_header: details.storage().get_header().to_bytes(),
-                                // Only include account code if the request did not contain it
-                                // (known by the caller)
-                                account_code: request_code_commitments
-                                    .contains(&details.code().commitment())
-                                    .not()
-                                    .then_some(details.code().to_bytes()),
-                            },
-                        )
-                    })
-                })
-                .collect()
+            let mut headers_map = BTreeMap::new();
+
+            for account_info in infos {
+                if let Some(details) = account_info.details {
+                    let request =
+                        match account_requests.iter().find(|req| req.account_id == details.id()) {
+                            Some(r) => r,
+                            None => panic!(),
+                        };
+
+                    let mut storage_slot_map_keys = Vec::new();
+
+                    for StorageMapKeysProof { storage_index, storage_keys } in
+                        &request.storage_requests
+                    {
+                        if let Some(StorageSlot::Map(storage_map)) =
+                            details.storage().slots().get(*storage_index as usize)
+                        {
+                            for map_key in storage_keys {
+                                let proof = storage_map.open(map_key);
+
+                                let slot_map_key = StorageSlotMapProof {
+                                    storage_slot: *storage_index as u32,
+                                    smt_proof: proof.to_bytes(),
+                                };
+                                storage_slot_map_keys.push(slot_map_key);
+                            }
+                        }
+                        // TODO: else error?
```

Comment:
This occurs if the user specifies an unknown slot key? Do we distinguish between `None` and zero in Miden, or is an empty slot equivalent to zero?